pkg/api/norman/customization/aks/handler.go

```diff
@@ -154,6 +165,33 @@ func (h *handler) getCloudCredential(req *http.Request, cap *Capabilities, credI
 	cap.ClientID = string(cc.Data["azurecredentialConfig-clientId"])
 	cap.ClientSecret = string(cc.Data["azurecredentialConfig-clientSecret"])
 
+	if cap.TenantID == "" {
+		annotations := cc.GetAnnotations()
+		tenantAnno, timestamp := annotations[tenantIDAnnotation], annotations[tenantIDTimestampAnnotation]
+		if tenantAnno != "" && timestamp != "" {
+			parsedTime, err := time.Parse(time.RFC3339, timestamp)
+			if err != nil {
+				return httperror.ServerError.Status, err
+			}
+			if parsedTime.Add(tenantIDTimeout).After(time.Now().UTC()) {
+				cap.TenantID = tenantAnno
+			}
+		}
+		if cap.TenantID == "" {
+			cap.TenantID, err = azureutil.FindTenantID(context.Background(), azure.PublicCloud, string(cap.SubscriptionID))
+			if err != nil {
+				return httperror.ServerError.Status, err
+			}
+			cc.Annotations[tenantIDAnnotation] = cap.TenantID
+			cc.Annotations[tenantIDTimestampAnnotation] = time.Now().UTC().Format(time.RFC3339)
+			// Ignore errors when updating the secret object. If the secret cannot be updated
+			// (perhaps due to a conflict error), the tenant ID will be re-fetched next time
+			// the proxy request is made and an update will be attempted again.
+			_, err := h.secretClient.Update(cc)
+			logrus.Debugf("encountered error while updating secret, ignoring: %v", err)
+		}
+	}
+
```

Comment:
I think putting this in it's own function would help with the nesting/readability.

Reply:
done